### DISCUSSION TITLE
csc.exe and vbc.exe startup performance

### DISCUSSION THREAD

**Discussion by Jinhuafei** [dotnet/roslyn#12403] [General]
Are these two exe enabled with [Multicore JIT](https://blogs.msdn.microsoft.com/dotnet/2012/10/18/an-easy-solution-for-improving-app-launch-performance/)? If not, can we enable it to improve the startup performance?

**Comment by gafter** [comment]
@agocke @jaredpar Can you please help answer this question?

**Comment by jaredpar** [comment]
This is something we can look into.  Had a couple of initial concerns:
1. Documentation mentions not working with NGEN.  Depending on the installation these binaries may or may not be ngen'd.  
2. In general these binaries aren't used.  Majority of builds go through MSBuild which in turn uses VBCSCompiler which eliminates the JIT issue altogether.  Is there a scenario where that is not happening? 
3. These are desktop specific APIs while our binaries are meant to be portable.  It's possible this can be used in our desktop specific setups though.

Also we need to actually measure and see if that helps us here.

**Comment by Jinhuafei** [comment]
Microsoft.CodeDom.Providers.DotNetCompilerPlatform executes csc.exe and vbc.exe to compile web app code. In a hosting environment, developer has no way to ngen Roslyn related bits. Without some optimization, the code startup performance is very bad.

**Comment by jaredpar** [comment]
@Jinhuafei correct but the cost of csc / vbc in this scenario has been measured as inconsequential in the past.  Sure it has to JIT but only a very, very small part of the compiler.  In our measurements it hasn't shown up as significant.

**Reply by MRatnam** [reply]
Our performance is 14s for ~100 compilations due to JIT (on 2023 dev machine). When using OS installed csc.exe, it's about a 10th of that. HUGE perf difference so this is quite a costly issue for us as we would like C#6 support in csc.exe but the cost is extremely expensive.

**Reply by tannergooding** [reply]
Are you using the Roslyn Compiler Server (vbcscompiler), which can be enabled via `<UseSharedCompilation>true</UseSharedCompilation>` or .NET Core images that are R2R compatible?

**Reply by MRatnam** [reply]
We're using Microsoft.CodeDom.Providers.DotNetCompilerPlatform csc.exe. Instantiating a CSharpCodeProvider pointing to the Roslyn version of the C# compiler for C# 6 support and compiling dynamically created classes (basically an equation engine) using provider.CompileAssemblyFromSource(). I don't know how to do this using the vbcscompiler.exe and UseSharedCompilation... https://learn.microsoft.com/en-us/dotnet/api/microsoft.csharp.csharpcodeprovider?view=dotnet-plat-ext-8.0

**Comment by Jinhuafei** [comment]
Based on my unscientific testing on a 16 cores VM, the cold startup perf diff is considerably big.
without ngen: 15s
with ngen: 9s

Test app: VS2015 website project template

**Comment by jaredpar** [comment]
We've seen considerably smaller numbers than that when we tested.  

Note: another option is to ngen the binaries.

**Comment by TyOverby** [comment]
Since our shipping binaries spend a *very* small amount of time jitting, I think we should close this issue; there's no real benefit for us to jit on multiple cores.

**Comment by jaredpar** [comment]
@TyOverby our shipping binaries spend a small time JIT'ing because they are NGEN on installation.  This issue is concerning the non-NGEN case where JIT can be a dominating force.

**Comment by TyOverby** [comment]
@jaredpar In what scenarios do we ship non ngened images?

**Comment by tannergooding** [comment]
@TyOverby, NuGet packages (such as our toolset compilers) are probably the primary scenario.

**Comment by jaredpar** [comment]
@tannergooding is correct that it's NuGet distribution.  In particular it's how a set of ASP.Net projects ship the compiler to the web server.  This is the driving problem in this issue.

**Comment by sharwell** [comment]
VBCSCompiler (the compiler server) is the primary way to avoid csc and vbc startup costs, and it works even for NuGet distribution of the toolset. I would interested to see if the same strategy could be used for https://github.com/dotnet/roslyn/issues/12403#issuecomment-233454489.
